### DISCUSSION TITLE
ResetMotorDriver & 2 new commands

### DISCUSSION THREAD

**Discussion by JulianOrteil** [ponoor/step-series-support#5] [General]
Hi Kanta,

Working through the documentation, I noticed that `/resetMotorDriver` is kind of vague on what it does. I went to the datasheet of the driver itself and don't see a similar command except for `Reset Device` (9.2.15 if we have the same documentation from Google lol). Can you provide some clarification on what this command is doing if it isn't following 9.2.15?

Second, would you be willing to look into adding the following commands if they're possible:

1. `/resetDevice`: Resets the entire device. A programmatic version of physically pressing the RESET button.
2. `/enablePositionReport [motorID]`: Automatically sends updates about the position of a motor. Kind of unsure as to how to handle situations where the motor is taking hundreds, or thousands, of steps per second. Perhaps adding a second `[interval]` param to this command, which denotes the interval between updates in milliseconds, would be a good work around?

Thanks for the awesome work on this project!
Jules

**Comment by kanta** [comment]
Hi Jules,

Thanks for your comments!

In STEP400/800 firmware, all values written to the motor driver chips are also stored in the firmware (RAM on the MCU) too. 
What the `/resetMotorDriver` command does is :
1. Execute `ResetDevice` command on the motor driver chip (L6470/PowerSTEP01).
2. Then set stored parameters to the driver chip. 

I used this command during a hardware debugging, since the early prototypes were made with long ribbon cables which cause SPI communication errors. So this command is a sort of leftover of developing process. Probably better to remove it from the documentation to keep things clear.

For new commands, both of them seems very useful!
I usually send `/getPosition` command periodically to acquire the current position but it would be nice to prepare the automatic report command. I will get to work on it soon.
Thank you for the suggestion!

**Reply by JulianOrteil** [reply]
Awesome, thanks in advance for looking into my requests and for your clarification!

Please let me know if there is anything I can do to help.
Jules

**Comment by kanta** [comment]
I'm implementing these new commands.
Basically I followed the syntaxes you suggested. `/resetMotorDriver` is very simple but others are;

## `/enablePositionReport [motorID] [interval]` 
Basically `interval` specifies the report period but when the value is set to 0, the reporting is disabled. This is probably better than adding one more bool arguments for enabling/disabling the function, and this style is already implemented in [`/setGoUntilTimeout`](https://ponoor.com/en/docs/step-series/osc-command-reference/homing/#setgountiltimeout_intmotorid_inttimeout) and [`/setReleaseSwTimeout`](https://ponoor.com/en/docs/step-series/osc-command-reference/homing/#setreleaseswtimeout_intmotorid_inttimeout).
The response is same as [`/getPosition`](https://ponoor.com/en/docs/step-series/osc-command-reference/position-management/#getposition_intmotorid).

I also added some more functions,

## `/enablePositionListReport [interval]` 
This command returns the position of all motors at once, combined to one OSC message. The response is 
`/positionList [motor#1 position] [motor#2 position] ... [motor#4 position]`
This command will be useful when you want to get all motor's positions, and same kind of syntax is already used in [`/setTargetPositionList`](https://ponoor.com/en/docs/step-series/osc-command-reference/servo-mode/#settargetpositionlist_intposition1_intposition2_intposition3_intposition4), though the data direction is opposite (device to host).

## `/getPositionList`
For polling style to get the position list.

---
Let me know if you find any better way, and I will proceed to the configTool update if it seems ok!

**Reply by JulianOrteil** [reply]
Everything here looks perfect, thanks again for the quick work!

Let me know when the firmware is ready and I'll update my driver board and try these out.

**Comment by kanta** [comment]
I updated the firmware both of STEP400 and STEP800.
https://github.com/ponoor/STEP400
https://github.com/ponoor/STEP800
New commands are;
- `/resetDevice`
- `/getPositionList`
- `/setPositionReportInterval`
changed the command name from `/enablePositionReport`
- `/setPositionListReportInterval`
changed the command name from `/enablePositionListReport`

configuration tools are also updated. 
Please check them out and please let me know if there are any problems!

**Reply by JulianOrteil** [reply]
Awesome, thanks for the work kanta. I will get testing right away!

**Comment by halpb** [comment]
Just saw this, this is going to be a lifesaver on my MAX console since I'm not polling/requesting /getPosition every 200ms!

**Reply by kanta** [reply]
I'm glad it seems to be useful! And thanks to Julian for the idea.

To use these functions, please do not forget to update the firmware to the latest one. You can find an explanation here;
https://ponoor.com/en/docs/step-series/technical-information/compile-the-arduino-sketch/
